RFC: Transaction type

samuel-rufi opened the pull request:
[Rendered](https://github.com/rufsam/bee-rfcs/blob/master/text/0000-bee-model-crate/0000-bee-model-crate.md)

Based on https://github.com/iotaledger/bee/issues/43

samuel-rufi committed:
Added RFC for implementing model crate

samuel-rufi committed:
Aligned with RFC folder structure. Little changes in code.

samuel-rufi committed:
Updating RFC with Bundle and Bundle Builder.

tsvisabo commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+- **branch hash** the hash of the second transaction referenced/approved = 81 trytes
+- **signature_and_message_fragment** contains the signature of the transfer or user-defined message data = 2187 trytes
+- **value** the transferred amount in IOTA  = 27 trytes
+- **address** receipt (output) address if value > 0, or withdrawal (input) address if value < 0 = 81 trytes
```
withdrawal => sender

Alex6323 replied:
I agree with replacing `receipt`, `withdrawal` with `receiver`, `sender`.

tsvisabo commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+A bundle then represents the collection of these transactions. It should be noted, that each transaction of a bundle will be sent separately through the network.
+
+Each transaction is identified by its hash, the transaction hash. Transactions should be final and therefore not modifiable after construction.
+The same applies to bundles. Bundles are final, and therefore not modifiable after construction.
```
What about transaction's metadata? this data is not gossiped of course, but nodes need it to check/update solidity/snapshot index

Alex6323 replied:
I would suggest to not mix `transaction` and `metadata` within one struct. A transaction itself is an immutable object which should be reflected in the code. The metadata for a transaction however can change i.e. be mutated/updated, like for example from how many neighbors this transaction was received. That's why I suggest having two structs here:
`Transaction` (immutable) and `TransactionMetadata` (mutable).
Sidenote: a `Transaction` itself is also *NOT* a vertex in the Tangle. It is rather a containment relationship: `Vertex` contains reference to a `Transaction`, and to at most two other vertices in the graph.

samuel-rufi replied:
Agree @Alex6323, we should separate `metadata` from `transaction`. We don't want to create a [god object](https://en.wikipedia.org/wiki/God_object).

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+
+# Summary
+
+This feature is responsible for the creation and interpretation of transactions and bundles.
```
I think I'ld like the `Summary` a little bit more detailed, what is actually introduced by this feature. I know that transactions and bundles are absolutely essential to IOTA, but for consistency with other RFCs could you write this section "as if" you would introduce those concepts with this proposal. I think we can expect the reader to have "some" DLT knowledge (so this summary should still remain rather short), but IMHO we should write those RFCs also for people that are fairly new to IOTA to get as many people involved as possible. The summary/abstract will be the first thing someone interested in it will read, and most of the times the last thing as well.

SuperFluffy replied:
I prefer language such as:

> This feature proposes the `Transaction` type to translate an incoming IOTA IRI transaction into a Rust data model.

Also, transactions and bundles seem to be two different concerns, operating at different levels of abstractions. I propose opening a separate RFC for transactions.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+IOTA is a distributed ledger that was designed for payment settlement and data transfer between machines and devices in the Internet of Things (IoT) ecosystem.
+The data packets that are sent through the network are called "transactions".
+Settlements or data transfers can be done with the help of these transactions. Payment settlements require the help of multiple transactions.
+
```
I think those sentences should move into the summary as they very well provide the necessary context to understand this proposal. To not repeat yourself in this section, you could pick up from the summary and elaborate on what transactions and bundles are in more detail, and how they are useful or even indispensable for the Bee framework.

SuperFluffy replied:
I agree with @Alex6323 here.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+
+## General
+
+A transaction consists of several fields (e.g. address, value, timestamp, tag). Each field is of static length. One transaction consists of 2673 trytes:
```
You keep this somewhat inconcrete here. Is this on purpose because of future changes to the transaction design? If it isn't the case then I would just write:
An IOTA transaction on the mainnet at the time of this proposal consists of 14(?) fields, ....
Instead of 'static length' I would prefer the term 'constant length', but that's just my preference.
Additionally to mentioning trytes, I would also add the trits and bytes count, because I would have the people in mind, that are not familiar with trinary. Let's not forget that most people think in binary.

SuperFluffy replied:
I also agree with @Alex6323 on this one. Ideally, we would be able to refer to some kind of document, spec, or just an SHA hash (of a github commit) where this message format is introduced. In the absence of that, one could just say that one wants to implement the message protocol as used by IRI (and cIRI?) in their most recent releases (reference the release, the git tag, and/or an SHA hash here).


Regarding the ternary encoding: I think it's fine mentioning that the incoming transaction is encoded in ternary, but for all intents and purposes it is not material what the specific encoding is. So, the fields of the `Transaction` `struct` should either be proper parsed Rust types, or they should be left in whichever encoding they are in when coming over the wire, but just extracted from the general message.

samuel-rufi replied:
To add the bytes count, we need to sort the encoding topic. I will add trits count. "static length" is replaced by now, also the transaction is defined in more detail. Also the field types are replaced. Will look the for the SHA hash of the tx layout were it was changed last time.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+
+Each transaction is identified by its hash, the transaction hash. Transactions should be final and therefore not modifiable after construction.
+The same applies to bundles. Bundles are final, and therefore not modifiable after construction.
+Having bundles as well as transactions final helps achieve correctness among the data objects.
```
I don't understand this sentence. How is 'correctness among data objects' defined? I have an idea of what you mean here, but can you make this more clear so someone from the outside can follow you here?

samuel-rufi replied:
I'm talking in the sense of software quality, as defined here:
"Correctness: The degree to which a system is free from [defects] in its specification, design, and implementation." [https://accu.org/index.php/journals/1585](https://accu.org/index.php/journals/1585)
Objects that contain properties/features that shouldn't be there. It's not meant in the sense of functional correctness regarding input-output behavior of an algorithm.
Will paraphrase it.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+
+Transactions can be built from Transaction Builders.
+Bundles can be built from Bundle Builders. Both builder objects can be manipulated as desired.
+It should be noted, Bundle Builders don't process Transactions (as transactions are final), instead, they process Transaction Builders.
```
what do you mean by 'they process transaction builder'? Aren't transaction builders used to simplify building transactions? Why would they needed to be processed. Maybe jus a wording thing, but I have different associations with the verb 'to process'.

samuel-rufi replied:
> Why would they needed to be processed.
Solved in [https://github.com/iotaledger/bee-rfcs/pull/3#discussion_r329781471](https://github.com/iotaledger/bee-rfcs/pull/3#discussion_r329781471)

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,279 @@
+    attachment_timestamp_upper_bound: i64,
+
+}
+```
````
I suggest a different approach here, because I see two main issues:
1. the transaction fields are (by design) of constant size known upfront. We should right of the bat make use of this knowledge, and store those fields in fixed sized arrays (if we go that route which I wouldn't suggest as it involves lots of allocations -> see proposal below)
2. this proposal of a transaction seems to be used for eagerly deserializing the transaction data from the received gossip (T5B1 or T9B2) bytes, which probably involves unnecessary calling of conversion functions in the hot-path, e.g. converting timestamps (i64). I haven't implemented a lot of IOTA nodes, but wouldn't it make more sense to only deserialize transaction fields lazily, that is on-demand? This might be more painful with T5B1 encoding, but with T9B2 each field has an integer byte size as well, which makes partial deserializing a breeze. That's why maybe it would be better to model `Transaction like so:
```Rust
struct Transaction {
    raw: Trits<T9B2>,
}
impl Transaction {
    ...
    fn address(&self) -> String {/*deserialize at specific location in self.raw*/}
    fn value(&self) -> i64 {/*same*/.}
    ...
}
``` 
It is also possible to choose a route in the middle of both extremes, where fields, that are always required, are deserialized eagerly. Just add fields for those to the `Transaction` struct (e.g. trunk, branch). If those T9B2 encoded trits are also used for transport (e.g. among Bee nodes) then in-memory copy operations could be reduced by a lot.

samuel-rufi replied:
Agree you fully, let's wait for the others what they think.

SuperFluffy replied:
If we indeed know the size of each field a priori, then we should model it accordingly. In addition, I think we should probably introduce a type for each of these fields in case they have to uphold invariants themselves.

Also, an array of `T`s with `m` elements, `[T; m]`, is allocated on the stack in Rust. If we want to have vec semantics, there is also the `smallvec` crate written by the Servo project.

As regards the `T9B2` encoding, we have discussed this yesterday (so after Alex left his review comment here) that we will for now stay within the already existing binary-coded trinary representations.

wusyong replied:
I'm also agree this. The return value can be what @SuperFluffy said, a fixed size array, or I think maybe it's also possible to be a reference slice.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,279 @@
+```
+
+As mentioned, Transactions are final. Transaction fields therefore should be only accessible by getter functions.
+Besides that, Transactions can be only built by Transaction Builders, or by the bytes of a received transaction.
````
Hopefully I don't come across as too nit-picky, but that sentence seems a bit vague to me. Suggestion:
"transactions can only be created from a transaction builder, or from a constructor function which expects encoded bytes as received e.g. from a network socket."

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,279 @@
+```rust
+impl Transaction {
+
+    pub fn from_transaction_builder(transaction_builder: &TransactionBuilder) -> Result<Transaction, TransactionBuilderValidationError> {
````
Suggestion: Do not let the `Transaction` struct know about a `TransactionBuilder`. There are two reasons:
1. Semantically: It's as if a car knows about its fabrication facility, but ideally a car shouldn't care about where it's produced. In the hierarchy of objects the `Transaction` is clearly subordinate to a `TransactionBuilder` which handles/operates on many transactions.
2. Breaking encapsulation/Dependency hell: IMHO one should only very carefully make types known to other types, and prevent it at all costs if possible. Refactoring code in one place should ideally not require changes in other parts.

samuel-rufi replied:
What we can do is moving initialization of the `Transaction` struct directly in the `build(self)` of the `TransactionBuilder`.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,279 @@
+```rust
+struct Transaction {
+
+    transaction_hash: String,
````
the hash of a transaction can be considered metadata of a `Transaction` and should not be part of the `Transaction` struct which itself is an immutable object.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+            Err(e) => { return Err(e) }
+        }
+
+        let (transaction_hash, nonce, timestamp) = Pow::compute(&transaction_builder);
```
IMHO this should be designed in a way that `Pow::compute(...)` returns a tuple, in this case a pair `(transaction, metadata)` where metadata contains the `transaction_hash` and the `timestamp`.

samuel-rufi replied:
Timestamp should be part of the transaction not metadata. `(transaction, metadata)` would imply, you would like to do the creation of a `Transaction` in `Pow`.

samuel-rufi replied:
Btw, if we do the creation of `Transaction` in Pow, we need to declare the Transaction fields as public. Else we end in something like this [https://doc.rust-lang.org/error-index.html#E0451](https://doc.rust-lang.org/error-index.html#E0451)
@Alex6323 moved the actual transaction struct initialization into Pow

Alex6323 replied:
@rufsam yea, your're right. I don't know why I wrote the timestamp would be metadata. I mean it is kind of, right?, but as it's also part of the transaction itself, of course it would be part of it.

samuel-rufi replied:
I'm not sure if we need an additional timestamp in the metadata. I think we are fine with having it only in the transaction.

Alex6323 commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,279 @@
+            attachment_timestamp: transaction_builder.attachment_timestamp.clone(),
+            attachment_timestamp_lower_bound: transaction_builder.attachment_timestamp_lower_bound.clone(),
+            attachment_timestamp_upper_bound: transaction_builder.attachment_timestamp_upper_bound.clone(),
+        })
```
I think this way of creating the transaction is dangerous. I think a TransactionBuilder should be destroyed having a method `build(self) -> Transaction` moving the internal state (the transaction to be built) out. I seems like what you want to have is a `TransactionFactory`, that has some template and creates clones off of it. But since transactions are completely unique instances where nothing can be really shared across them, the factory pattern doesn't seem to be a good choice here.

samuel-rufi replied:
Thanks for the comment. Agree, moving ownership and destroy it afterwards seems to be the better choice.

samuel-rufi committed:
Replaced "receipt" with "receiver" and "withdrawal" with sender.
Updated the sentence how transactions can be built.
Removed tx hash from Transaction struct.

samuel-rufi committed:
Added metadata struct, and a proof of work example

samuel-rufi committed:
Snapshot index fix

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+- **nonce** is the Proof-of-Work nonce of the transaction = 27 trytes
+- **bundle hash** is the hash of the entire bundle = 81 trytes
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
```
I am wondering if we can make a nice ASCII art out of this.

thibault-martinez replied:
What about a table Name | Description |  Size ?

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
+
+A bundle is a collection of specific transactions. Bundles are required to bundle related information.
```
I think it's fine to mention bundles, but they should probably be in their own RFC to keep the focus on the transaction itself.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+    }
+
+}
+```
````
I think you can safely remove most of the getter functions above to demonstrate what you mean. :) Let's mention that the fields should be left private and immutable.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+}
+```
+
+The Transcation_Builder struct contains public accessible fields. 
````
`Transcation_Builder -> TransactionBuilder`

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+    pub attachment_timestamp: i64,
+    pub attachment_timestamp_lower_bound: i64,
+    pub attachment_timestamp_upper_bound: i64,
+}
```
For a transaction builder, the proper semantics are to keep the fields as `Option<T>`. Also, I would also leave the fields `non-pub` and only allow setting them through setter methods.

samuel-rufi replied:
> For a transaction builder, the proper semantics are to keep the fields as Option<T>

Great idea

> Also, I would also leave the fields non-pub and only allow setting them through setter methods.

Where would you put the validation of the set values? Directly in the setter, or as it currently is in `build()`? Currently the validation happens one time, else it could be done in each setter directly. Not sure what suites best here. If we decide in each setter, we need to keep in mind that every setter could throw an error and how handy it is for the programmer.

samuel-rufi replied:
Regarding `non-pub` there is the idea to build the actual Transaction (creating the struct) in Pow::compute(). Currently we have the Transaction fields `non-pub`. It seems like this ends in something like this https://doc.rust-lang.org/error-index.html#E0451

wusyong replied:
Let's use visibility we introduced yesterday then. I think `pub(crate)` should be fine.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+    pub is_solid: bool,
+    pub usize: snapshotIndex
+}
+```
````
As before, I would leave this `non-pub`. Since we are probably interested in upholding invariants, we shouldn't give out direct access to the fields.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+```rust
+pub struct Pow;
+
+impl Pow {
````
Let's move all this into its own section and call it “how this will be used” or something like that.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+```
+
+
+### Bundle
````
This should go into its own RFC.

thibault-martinez replied:
Yeah not sure we  should mention PoW at all here.

samuel-rufi replied:
There was an idea to build the Transaction directly in the compute of PoW. If we go with it, IMO we need to show what role PoW has and how the Transaction gets finally built.

samuel-rufi replied:
@SuperFluffy Agree

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+
+# Drawbacks
+
+Without any bee-model crate, nodes can not exchange transactions. Therefore this crate seems necessary.
```
Drawbacks should ideally also discuss the drawbacks of this design. :)

I am thinking of the `String` types, for example, which could also be done as their own types.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+
+Without any bee-model crate, nodes can not exchange transactions. Therefore this crate seems necessary.
+
+# Rationale and alternatives
```
An additional rationale is definitely that this kind of interface is relatively minimal and easily extended upon in a future iteration.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+- The distinction between Transaction and Transaction Builder as well as Bundle and Bundle Builder makes the code cleaner 
+and helps achieve correctness among the data objects. Properties are clearly assigned to specific data objects and not mixed up.
+
+- The proposed crate interface is intuitive. Completely different alternatives did not naturally come to mind.
```
If it's intuitive is for users to decide. I agree however that it is a very natural proposition to basically treat the incoming transaction as some encoded binary blob.

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+
+# Unresolved questions
+
+- How is Proof-Of-Work acceleration handled?
```
This should probably be further explained. Is `Proof-of-Work acceleration` really material to this proposal? If not, I would leave it out.

Important unresolved questions:

+ How to handle deserialization of the incoming, encoded transaction?

SuperFluffy commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
++ Feature Name: `bee-model-crate`
```
We should rename it to something like `iri-transaction-model` or `mainnet-transaction-model` or `v1-transaction-model`.

thibault-martinez replied:
I'd vote for versioning since `mainnet-transaction-model` is not time-proof and `iri-transaction-model` is not really the case.

samuel-rufi replied:
How do you want to version it?

SuperFluffy reviewed (changes requested):
I like this RFC because I understand its purpose, and because I feel like it fulfills (or is very close to fulfilling) the requirement of taking care of a specific, cleanly separated concern.

The proposal should definitely separate the `Transaction` and `Bundle` types into their respective RFCs, because these seem to be sufficiently different.

The RFC can mention that the `Transaction` type could live in its own crate.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
++ Feature Name: `bee-model-crate`
```
```suggestion
+ Feature name: `bee-model-crate`
```
The template changed since you opened.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
++ Feature Name: `bee-model-crate`
++ Start Date: 2019-09-06
```
```suggestion
+ Start date: 2019-09-06
```
The template changed since you opened.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
++ Feature Name: `bee-model-crate`
++ Start Date: 2019-09-06
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)
```
```suggestion
+ RFC PR: [iotaledger/bee-rfcs#3](https://github.com/iotaledger/bee-rfcs/pull/3)
```

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+
+- **trunk hash** the hash of the first transaction referenced/approved = 81 trytes
+- **branch hash** the hash of the second transaction referenced/approved = 81 trytes
+- **signature_and_message_fragment** contains the signature of the transfer or user-defined message data = 2187 trytes
```
`signature_or_message_fragment` ?

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+- **nonce** is the Proof-of-Work nonce of the transaction = 27 trytes
+- **bundle hash** is the hash of the entire bundle = 81 trytes
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
```
I know I'm being picky here but can you list  them in their actual order ? I really think it helps a lot.

samuel-rufi replied:
How do you want to order them?

thibault-martinez replied:
Order in which the fields appear in a transaction.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+- **nonce** is the Proof-of-Work nonce of the transaction = 27 trytes
+- **bundle hash** is the hash of the entire bundle = 81 trytes
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
```
That is not true. The total number of transactions in the bundle is `last_index + 1`.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,334 @@
+```rust
+struct Transaction {
+
+    signature_fragments: String,
````
I'd like the name to reflect that it doesn't necessarily contain a signature. `signature_or_message` ?

wusyong replied:
Is it possible to even not contain this field if the instance doesn't have `signature_or_message`? It would be sth like this:
```suggestion
    signature_or_message: Option<String>,
```

thibault-martinez replied:
`Option` still allocates memory for the field so it's not really helpful

samuel-rufi replied:
@SuperFluffy what are your thoughts regarding `Option`. Semantically, it would be more correct, right?

thibault-martinez replied:
I'm not sure it makes more sense "semantically", an empty sig_or_msg doesn't mean there is no sig_or_msg. There is always a sig_or_msg :)

samuel-rufi replied:
Agree, so maybe Option isn't necessary. And if we use static types, we could initialize it with an empty (9's) `SignatureOrMessageFragment` and could be sure the internal state of the builder object is always valid.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+    attachment_timestamp_lower_bound: i64,
+    attachment_timestamp_upper_bound: i64,
+
+}
```
Still being picky, can you order the fields ?

samuel-rufi replied:
How do you want to order them?

thibault-martinez replied:
Order in which the fields appear in a transaction.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+    tag: String,
+    attachment_timestamp: i64,
+    attachment_timestamp_lower_bound: i64,
+    attachment_timestamp_upper_bound: i64,
```
Timestamps are unsigned

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,334 @@
+    pub transaction_hash: String,
+    pub is_solid: bool,
+    pub snapshot_index: usize
+}
```
I think `TransactionMetadata` will be the topic of another RFC so I wouldn't necessarily put the structure here. Or at least not `is_solid` and `snapshot_index`.

samuel-rufi committed:
Update "Name" to "name" since template change

Co-Authored-By: Thibault Martinez <thibault.martinez.30@gmail.com>

samuel-rufi committed:
Update "Date" to "date" since template change

Co-Authored-By: Thibault Martinez <thibault.martinez.30@gmail.com>

samuel-rufi committed:
update iotaledger/bee-rfcs#000 to iotaledger/bee-rfcs#3

Co-Authored-By: Thibault Martinez <thibault.martinez.30@gmail.com>

samuel-rufi committed:
Order fields, fix i64 to u64, updated "Unresolved questions"

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,290 @@
+
+    signature_fragments: String,
+    address: String,
+    value: u64,
```
```suggestion
    value: i64,
```

samuel-rufi committed:
Fix description of last_index field. Fix unsigned/signed types

samuel-rufi committed:
Moved motivation part to summary, displayed transaction fields in a table, added unresolved question, removed Bundle part since its better to have it as separate RFC.

samuel-rufi committed:
Enlarge RFC, Adding questions to "Unresolved questions"

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
++ Feature name: `bee-transaction-crate`
```
Not a full crate

samuel-rufi replied:
lets call it module?

wusyong replied:
How about this?
```suggestion
+ Feature name: `transaction-module`
```

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+| **current_index**  | the position of the transaction in its bundle  | 9 trytes |
+| **last_index**  | the index of the last transaction in the bundle  | 9 trytes |
+| **bundle hash**  | the hash of the bundle essence | 81 trytes |
+| **trunk hash**  | the hash of the first transaction referenced/approved | 81 trytes |
```
```suggestion
| **trunk**  | the hash of the first transaction referenced/approved | 81 trytes |
```

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+| **last_index**  | the index of the last transaction in the bundle  | 9 trytes |
+| **bundle hash**  | the hash of the bundle essence | 81 trytes |
+| **trunk hash**  | the hash of the first transaction referenced/approved | 81 trytes |
+| **branch hash**  | the hash of the second transaction referenced/approved | 81 trytes |
```
```suggestion
| **branch**  | the hash of the second transaction referenced/approved | 81 trytes |
```

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+| **bundle hash**  | the hash of the bundle essence | 81 trytes |
+| **trunk hash**  | the hash of the first transaction referenced/approved | 81 trytes |
+| **branch hash**  | the hash of the second transaction referenced/approved | 81 trytes |
+| **nonce**  | is the Proof-of-Work nonce of the transaction | 27 trytes |
```
Sorry to bother again with this but `nonce` is the last field :p

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+| **attachment_timestamp_upperbound**  |  *not specified* | 9 trytes |
+
+As reflected in the table, each field is of constant length. Totally, one transaction consists of **2673 trytes**.
+Transactions should not be **modifiable** after construction. Transaction fields therefore only should be accessible by getter functions.
```
Can you expand a little bit on why it should not be modifiable ?  We can't just say it should not.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+    bundle_hash: String,
+    trunk: String,
+    branch: String,
+    nonce: String,
```
`nonce` is the last field.

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+| **timestamp**  | the time when the transaction was issued  | 9 trytes |
+| **current_index**  | the position of the transaction in its bundle  | 9 trytes |
+| **last_index**  | the index of the last transaction in the bundle  | 9 trytes |
+| **bundle hash**  | the hash of the bundle essence | 81 trytes |
```
```suggestion
| **bundle**  | the hash of the bundle essence | 81 trytes |
```

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+    timestamp: u64,
+    current_index: usize,
+    last_index: usize,
+    bundle_hash: String,
```
I'd just call that `bundle`

samuel-rufi replied:
Then we probably also should call `trunk hash` `trunk` and `branch hash` `branch`

thibault-martinez replied:
They are already called `trunk` and `branch`

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,264 @@
+
+### TransactionBuilder
+
+The TransactionBuilder offers a **simple and convenient** way to built transactions. It's defined as follows:
```
```suggestion
The TransactionBuilder offers a **simple and convenient** way to build transactions. It's defined as follows:
```

samuel-rufi committed:
changed "built" do "build"

Co-Authored-By: Thibault Martinez <thibault.martinez.30@gmail.com>

samuel-rufi committed:
removed "hash suffix" in table. moved nonce after all other fields

samuel-rufi committed:
Merge remote-tracking branch 'origin/master'

lunfardo314 commented:
`Transaction` and related types better be abstract traits since the beginning.
I refer to requirements of Q-nodes to Bee: https://github.com/iotaledger/bee-rfcs/issues/6#issue-494583275 . 
The assemblies would need multiple types of transactions in the same bundle and some new constrains on the ledger, other than usual constraints on token accounts. For example, there's a need to support unforkable `chains`, signed by the assembly.

samuel-rufi committed:
Update text/0000-bee-model-crate/0000-bee-model-crate.md

Co-Authored-By: Wu Yu Wei <wusyong9104@gmail.com>

samuel-rufi committed:
Replaced types of Transaction/TransactionBuilder. Fields-validator isn't needed anymore, removed it from the build() of TransactionBuilder. Changed visibility of fields in TransactionBuilder to private. Setter/getter functions will be used to set/access fields.
Moved from_bytes() constructor of Transaction to TransactionBuilder. TransactionBuilder also allows to set Timestamp and Nonce manually in case they are available.

samuel-rufi committed:
Changed types of TransactionMetadata and provided impl with setter & getter.

samuel-rufi committed:
Merge remote-tracking branch 'origin/master'

thibault-martinez commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -146,7 +170,7 @@ Each transaction contains a mutable metadata which is defined as following:
 ```rust
 pub struct TransactionMetadata {
-    pub transaction_hash: String,
+    pub transaction_hash: TransactinHash,
````
Typo

samuel-rufi committed:
added why transactions should not be modifiable, what is meant with "transaction is a tuple", and fixed typo

samuel-rufi committed:
Update unresolved questions.

samuel-rufi committed:
Add explanation for the field types. Added unresolved questions. Added info regarding the Pow:compute(&TransactionBuilder) call since it should not be overseen and reviewed.

wusyong commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -299,8 +296,6 @@ Without any bee-model crate, nodes can not exchange transactions. Therefore this
+- Should we introduce a fluent API to the TransactionBuilder?
 
-- Introducing a fluent API to the TransactionBuilder?
+- Should we use Option<T> for the fields in TransactionBuilder? 
```
Some fields would need this I guess. Others like timestamp, index etc are mandatory.

SuperFluffy replied:
The `build` method is responsible for ensuring that mandatory fields are set, returning `Result::Err` if not. The type of the field is `Option` if there is no natural (read: meaningful) way of providing a default.

Commit:
Simplified summary, use transaction version following IRI

Commit:
Added motivational section

Commit:
Rewrote detailed design

Commit:
Fix table formatting

Commit:
Simplify Transaction design

wusyong commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,277 @@
+
+| Name                              | Description                                            | Size (in trits) |
+| ---                               | ---                                                    | ---             |
+| `signature_or_message_fragment`   | contains the signature of the transfer                 |                 |
+|                                   | or user-defined message data                           | 6561            |
```
I think it's no need to be separate line. Besides, this would become totally different row. Normal markdown display would just adjust them.

```suggestion
| `signature_or_message_fragment` | contains the signature of the transfer or user-defined message data | 6561 |
```

wusyong replied:
@SuperFluffy I think we could ignore max chars limit during table format. We should make representation correct at least.

wusyong commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,277 @@
+| ---                               | ---                                                    | ---             |
+| `signature_or_message_fragment`   | contains the signature of the transfer                 |                 |
+|                                   | or user-defined message data                           | 6561            |
+| `address`                         | receiver (output) if value > 0,
+|                                   | or sender (input) if value < 0                         | 243             |
```
ditto

SuperFluffy commented:
Yeah, I wanted to stay within the number of max chars per line we discussed.

I thought GitHub markdown supports multiline table cells, but I either
chose the wrong table format or GitHub doesn't support these kinds of raw
tables as pandoc does.

On Mon, Oct 14, 2019, 4:30 PM Wu Yu Wei <notifications@github.com> wrote:

> *@wusyong* commented on this pull request.
> ------------------------------
>
> In text/0000-bee-model-crate/0000-bee-model-crate.md
> <https://github.com/iotaledger/bee-rfcs/pull/3#discussion_r334510886>:
>
> > +| `signature_or_message_fragment`   | contains the signature of the transfer                 |                 |
>
> +|                                   | or user-defined message data                           | 6561            |
>
>
> I think it's no need to be separate line. Besides, this would become
> totally different row. Normal markdown display would just adjust them.
> ⬇️ Suggested change
>
> -|                                   | or user-defined message data                           | 6561            |
>
> +| `signature_or_message_fragment` | contains the signature of the transfer or user-defined message data | 6561 |
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/iotaledger/bee-rfcs/pull/3?email_source=notifications&email_token=AAFLF6PY3XCAV3PIT2664J3QOR67NA5CNFSM4IUNFOQKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCH3CK2I#pullrequestreview-301344105>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFLF6PUP6JTAXDSMFLQ2N3QOR67NANCNFSM4IUNFOQA>
> .
>

wusyong commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
````diff
@@ -0,0 +1,277 @@
+The `from_slice` and `from_reader` methods will use a `TransactionDraft` internally,
+but these are implementation details.
+
+The `TransactionError` is not fully fleshed out yet. For the time being, it
+definitely contains `io::Error`:
+
+```rust
+pub enum TransactionError {
+    Io(io::Error),
+}
````
I think we should also start considering which error handling library to use. I saw there's another new one called [thiserror](https://github.com/dtolnay/thiserror). What do you think?

wusyong commented on text/0000-bee-model-crate/0000-bee-model-crate.md:
```diff
@@ -0,0 +1,277 @@
+        unimplemented!()
+    }
+
+    pub fn address(&mut self, address: Address) -> &mut Self {
```
How about give a proper setter method prefix like `set_address`?

samuel-rufi replied:
Yes, as RFC #344 notes it:
[https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis)

SuperFluffy commented:
Incorporated into #20.
